src/webgpu/api/validation/encoding/programmable/pipeline_bind_group_compat.spec.ts

```diff
@@ -141,3 +173,90 @@ g.test('it_is_invalid_to_draw_in_a_render_pass_with_missing_bind_groups')
+      { setBindGroup1: true, setBindGroup2: false, setUnusedBindGroup: true, _success: false },
+      { setBindGroup1: false, setBindGroup2: true, setUnusedBindGroup: true, _success: false },
+      { setBindGroup1: false, setBindGroup2: false, setUnusedBindGroup: false, _success: false },
+    ])
+  )
+  .unimplemented();
+
+g.test('binding_mismatch_in_bgl_in_render_pass')
+  .desc(
+    'Tests the binding number must exist or not exist in both bindGroups[i].layout and pipelineLayout.bgls[i]'
+  )
+  .paramsSubcasesOnly(u =>
+    u.combine('cmd', kDrawCmds).combineWithParams([
+      { bgBindings: [0, 1, 2] as const, plBindings: [0, 1, 2] as const, _success: true },
+      { bgBindings: [0, 1, 2] as const, plBindings: [0, 1, 3] as const, _success: false },
+      { bgBindings: [0, 2] as const, plBindings: [0, 2] as const, _success: true },
+      { bgBindings: [0, 2] as const, plBindings: [2, 0] as const, _success: true },
+      { bgBindings: [0, 1, 2] as const, plBindings: [0, 1] as const, _success: false },
+      { bgBindings: [0, 1] as const, plBindings: [0, 1, 2] as const, _success: false },
+    ])
+  )
+  .unimplemented();
+
+g.test('binding_mismatch_in_bgl_in_compute_pass')
+  .desc(
+    'Tests the binding number must exist or not exist in both bindGroups[i].layout and pipelineLayout.bgls[i]'
+  )
+  .paramsSubcasesOnly(u =>
+    u.combine('cmd', kDispatchCmds).combineWithParams([
+      { bgBindings: [0, 1, 2] as const, plBindings: [0, 1, 2] as const, _success: true },
+      { bgBindings: [0, 1, 2] as const, plBindings: [0, 1, 3] as const, _success: false },
+      { bgBindings: [0, 2] as const, plBindings: [0, 2] as const, _success: true },
+      { bgBindings: [0, 2] as const, plBindings: [2, 0] as const, _success: true },
+      { bgBindings: [0, 1, 2] as const, plBindings: [0, 1] as const, _success: false },
+      { bgBindings: [0, 1] as const, plBindings: [0, 1, 2] as const, _success: false },
+    ])
+  )
+  .unimplemented();
+
+g.test('visibility_mismatch_in_bgl_in_render_pass')
+  .desc('Tests the visibility in bindGroups[i].layout and pipelineLayout.bgls[i] must be matched')
+  .paramsSubcasesOnly(u =>
+    u
+      .combine('cmd', kDrawCmds)
+      .combine('bgVisibility', kShaderStages)
+      .combine('plVisibility', [
+        GPUConst.ShaderStage.VERTEX,
+        GPUConst.ShaderStage.FRAGMENT,
+      ] as const)
+  )
+  .unimplemented();
+
+g.test('visibility_mismatch_in_bgl_in_compute_pass')
+  .desc('Tests the visibility in bindGroups[i].layout and pipelineLayout.bgls[i] must be matched')
+  .paramsSubcasesOnly(u => u.combine('cmd', kDispatchCmds).combine('bgVisibility', kShaderStages))
+  .unimplemented();
+
+g.test('resource_type_mismatch_in_bgl_in_render_pass')
+  .desc(
+    'Tests the binding resource type in bindGroups[i].layout and pipelineLayout.bgls[i] must be matched'
+  )
+  .paramsSubcasesOnly(u =>
+    u
+      .combine('cmd', kDrawCmds)
+      .combine('bgResource', kBindableResources)
```

Comment:
I think this doesn't cover all of the types. e.g. I don't find `external_texture`. Maybe we should list uncovered cases into TODO and add helpers for them

Reply:
Changed `kBindableResources` to `kResourceTypes` for all types checking. In this tests, we just check resource types mismatched, for same resource types, 'buffer' for example, we will not check the buffer binding type mismatched, which still kept in TODOs:
> the binding resource in bindGroups[i].layout is "group-equivalent" (value-equal) to pipelineLayout.bgls[i].